feature-cmake-object-libraries: BLE as OBJECT library

0xc0170 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

Main goal: mbed-ble to be OBJECT library. The changes should be simple. I created lot of small CMake targets but linked them to mbed-ble. It's out of scope of this feature branch to remove labels.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

@multiplemonomials @ladislas please review as well

0xc0170 committed:
CMake ble: set library as OBJECT

mbed-ble to be OBJECT library.

Cordio is interface as it is part of mbed-ble but as an external component that we fetch from upstream cordio repository.

0xc0170 committed:
CMake ble: fix libraries in drivers

0xc0170 committed:
CMake: add components

New utility to keep track of components and their libraries.

The idea for this is:

- keep valid names of components in the Mbed OS tree (to check if they exist and error if not)
- add additional libraries for each component

A target just invokes the function and adds extra lib required for a target.

0xc0170 committed:
CMake STM32: add extra libraries to mbed-ble

0xc0170 committed:
CMake Cypress: add extra libraries to mbed-ble

0xc0170 committed:
CMake targets: fix includes for components

0xc0170 committed:
CMake Ambiq Micro: add extra libraries to mbed-ble

0xc0170 committed:
CMake STM32WB: add extra libraries to mbed-ble

0xc0170 committed:
CMake NORDIC: add ble extra libraries

0xc0170 commented on connectivity/FEATURE_BLE/CMakeLists.txt:
```diff
@@ -1,11 +1,22 @@
+# Add drivers for BLE
+# Once we remove the old tools and their path requirements, we should move 
+# drivers to BLE stack to be in the same place
+add_subdirectory(../drivers/ble ${CMAKE_CURRENT_BINARY_DIR}/drivers/ble)
```
Not certain if this is done properly. If I should rather just use `${MBED_PATH}/connectivity/drivers/ble` ? 

These are drivers for ble (extensions) so they should be included here rather than in driver.

0xc0170 replied:
reverting this back to as it was (if we ever move drivers).

0xc0170 committed:
CMake ble: revert drivers inclusion

0xc0170 committed:
CMake ble drivers: remove linking core and ble - they are default for ble

0xc0170 committed:
CMake ble drivers: add EXCLUDE_FROM_ALL to optional drivers

hugueskamba commented on connectivity/drivers/CMakeLists.txt:
```diff
@@ -2,8 +2,8 @@
 add_subdirectory(802.15.4_RF)
-add_subdirectory(ble)
 add_subdirectory(cellular)
+add_subdirectory(ble)
```
Any reason why this was moved?

0xc0170 replied:
Wrong revert, I'll fix that

0xc0170 replied:
Done

hugueskamba commented on connectivity/drivers/ble/FEATURE_BLE/COMPONENT_BlueNRG_MS/CMakeLists.txt:
```diff
@@ -1,12 +1,9 @@
-    INTERFACE
-        BlueNrgMsHCIDriver.cpp
-)
+add_library(mbed-ble-bluenrg_ms INTERFACE)
```
How do we make the decision which library should be `INTERFACE`?

0xc0170 replied:
if CMake would allow circular deps, this would be OBJECT library.

This library provides functionality declared in BLE.
To overcome this limitation, the main library that is user facing lib (mbed-ble in this case) is OBJECT library. Extensions like components are INTERFACES as they can't built without the library that declares the functionality.

hugueskamba commented on connectivity/drivers/ble/FEATURE_BLE/TARGET_STM32WB/CMakeLists.txt:
```diff
@@ -1,13 +1,15 @@
 # SPDX-License-Identifier: Apache-2.0
 
-target_include_directories(mbed-ble
+add_library(mbed-ble-stm32wb INTERFACE)
```
Here for example, why isn't this library an `OBJECT` lib?

0xc0170 replied:
I'll add a comment to main CMake file about how drivers are attached to the mbed-ble for now to explain it in one place.

0xc0170 replied:
Done, also provided details in the commit message.

hugueskamba commented on connectivity/drivers/ble/FEATURE_BLE/COMPONENT_CYW43XXX/firmware/TARGET_CYW9P62S1_43012EVB_01/CMakeLists.txt:
```diff
@@ -1,7 +1,9 @@
 # SPDX-License-Identifier: Apache-2.0
 
-target_sources(mbed-ble
+add_library(mbed-ble-cyw9p62s1_43012evb_01 INTERFACE)
```
This CMake library is not referenced in any Mbed target. Should it?

0xc0170 replied:
Yes, this should be a new issue. I'll report it.

hugueskamba commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CYSBSYSKIT_01/CMakeLists.txt:
```diff
@@ -60,3 +61,4 @@ target_compile_definitions(mbed-cysbsyskit-01
 )
 
 mbed_post_build_psoc6_merge_hex("CYSBSYSKIT_01")
+mbed_add_component_libraries(mbed-ble mbed-ble-cyw43xxx-43012 mbed-ble-cyw43xxx mbed-ble-cyw43xxx-psoc6)
```
Should `mbed-ble-cyw43xxx-43012` and `mbed-ble-cyw43xxx` be here?

0xc0170 replied:
I'll fix, just one should do the trick if we got deps correct.

0xc0170 replied:
`mbed-ble-cyw43xxx` is not needed, the first yes - its' firmware. The layout of this is not trivial but seems like both must be defined by a target at the moment.

hugueskamba commented on tools/cmake/components.cmake:
```diff
@@ -0,0 +1,18 @@
+    # we use global symbols, using upper case
+    string(TOUPPER ${component} component_upper)
+    if(component IN_LIST MBED_COMPONENTS_LIST)
+        set(${component_upper}_EXTRA_LIBRARIES "${${component}_EXTRA_LIBRARIES};{ARGN}" CACHE INTERNAL "")
```
Can we pass the name of the extra library list instead of building it from `component`? I am even wondering if the function is needed at all...

rwalton-arm replied:
> I am even wondering if the function is needed at all...

I am also wondering why we need this function. If there's a strong justification for this we should document it somewhere, preferably in the commit message.

Another issue I'm having with this code is that we have now introduced this "component" terminology, which is confusing, as CMake does already have a "components" concept. I think the overloading of the word "component" here will be confusing for future developers.

0xc0170 replied:
I can fix wording, first lets agree if we want a variable to be accessible or function/macro.

We need to have a list of extra modules accessible in the tree. This can be done via cache variable or property. 

If it's cache, a user would need to append to it (I wish there is list append or similar for a cache, I only was able to   make ti work this way). Similar as it is done in the function:
```
set(MBED_BLE_EXTRA_LIBRARIES "${MBED_BLE_EXTRA_LIBRARIES};target-extra-libs" CACHE INTERNAL "")
```

Is that straightforward to a user  to do this ? If you both agree, I'll switch back.

hugueskamba replied:
Ok I think I understand that the point of this function is to be a sort of utility function that helps to add values to a cache variable. The use case is as follows:
If we do not have the function, for a tree like:
```
+ foo/
    + CMakeLists.txt
    + bar/
        + CMakeLists.txt
```
If `foo/CMakeLists.txt` creates the cache variable `XXX_EXTRA_LIBRARIES` and `bar/CMakeLists.txt` then wants to append to it it needs to do:
```
set(XXX_EXTRA_LIBRARIES
    ${XXX_EXTRA_LIBRARIES}
    <OTHER_CONNECTIVITY_DRIVER_CMAKE_TARGET>
    CACHE INTERNAL ""
)
```
This does not seem very elegant and I see how the function helps. I do not have a strong opinion on whether to keep the function or not. However, I would request for it to be generic as in it's just a function to append values to a cache variable. It should therefore have as input the name of the variable and the values to add to it.

hugueskamba replied:
However, not having the function gives the possibility to easily remove libraries added by parent directories. In the example above, if `bar/CMakeLists.txt` wanted to remove one of the libraries added it could just overwrite the cache variable `XXX_EXTRA_LIBRARIES` with `set(XXX_EXTRA_LIBRARIES
 <NOTHING_OR_ONLY_THE_DRIVERS_NEEDED> CACHE INTERNAL "")`.

0xc0170 replied:
I'll switch it back.

0xc0170 replied:
Fixed

rwalton-arm commented on targets/TARGET_Ambiq_Micro/CMakeLists.txt:
```diff
@@ -1,6 +1,8 @@
 # Copyright (c) 2020-2021 ARM Limited. All rights reserved.
 # SPDX-License-Identifier: Apache-2.0
 
+include(${MBED_PATH}/tools/cmake/components.cmake)
```
This was preexisting but, I think it's a poor architecture to to define this `MBED_PATH` variable in the application when mbed-os depends on it everywhere. CMake provides several variables to discover the current path, we should use those variables if we can. It seems very strange that we rely on the app to set this `MBED_PATH` variable, and it leads to a very confusing dependency on the app. That dependency makes building mbed-os more difficult than it should be in various cases. If the app needs to include something from mbed-os' CMake modules, it should probably define that path itself, in another variable mbed-os doesn't depend on. 

Another possible approach would be to append our module folder to [CMAKE_MODULE_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_MODULE_PATH.html) which may avoid the need to hardcode the module include paths everywhere.

0xc0170 replied:
I'll fix using CMAKE_MODULE_PATH but I guess it won't be needed if the previous question is resolved via removing this   .cmake file completely and providing just a variable to a user.

rwalton-arm replied:
My comment is more about the use of the `MBED_PATH` variable in general, rather than just in this new code. I would like to see us remove all usage of `MBED_PATH` from mbed-os, removing the dependency on the consumer defining this variable (currently we define it in the application `CMakeLists.txt` in the examples). That work should probably be a backlog item because we have roughly 280 usages of `MBED_PATH` in mbed-os at the moment.

ladislas replied:
I agree with @rwalton-arm idea of removing the need for the app to define `MBED_PATH` -- if it is needed internally, it can be renamed something like `MBED_INTERNAL_PATH` so that people won't define the same variable by accident.

> If the app needs to include something from mbed-os' CMake modules, it should probably define that path itself, in another variable mbed-os doesn't depend on.

Yep, I thing that's the way to go.

0xc0170 replied:
> My comment is more about the use of the MBED_PATH variable in general, rather than just in this new code. I would like to see us remove all usage of MBED_PATH from mbed-os, removing the dependency on the consumer defining this variable (currently we define it in the application CMakeLists.txt in the examples). That work should probably be a backlog item because we have roughly 280 usages of MBED_PATH in mbed-os at the moment.

+1, can you create an issue in this repo? It should be easy to fix.

0xc0170 committed:
CMake ble: remove component add function, use regular variable

use MBED_<library>_EXTRA_LIBRARIES variables we provide

0xc0170 committed:
CMake ble: revert the move

0xc0170 committed:
CMake ble: fix naming CMake libraries, using -

0xc0170 committed:
CMake CYW9P62S1: add one more ble driver

0xc0170 committed:
CMake ble: remove not completed add/remove libraries from targets

This needs much more time to be done properly. Use what we have currently to be able to
use mbed-ble as object library.

Once we get add/remove libraries from targets, fix should be easier as we already have
own CMake targets per driver, just connect them together with mbed-ble on request.

0xc0170 committed:
CMake targets: remove empty space

0xc0170 committed:
CMake ble: link drivers to ble

0xc0170 committed:
CMake ble: bring back labels

0xc0170 committed:
CMake ble: fix cordio-ll name

0xc0170 committed:
CMake ble: add a comment why drivers are INTERFACE lib

If it is an OBJECT it would need mbed-ble and mbed-core. As we want to link these drivers to
mbed-ble to be available for an app (the only required step for an app is to link to mbed-ble and
we figure out internally what is required for a specific board - using config library add/remove).

This can be refactored once we get CMake config story to replace our config (or at least targets.json information like add/remove libraries).

0xc0170 committed:
CMake connectivity: expose mbed-ble as drivers link to it

hugueskamba commented on connectivity/CMakeLists.txt:
```diff
@@ -3,10 +3,7 @@
 
 # List of all connectivity libraries available.
```
Please remove this comment.

0xc0170 replied:
removed

hugueskamba commented on connectivity/FEATURE_BLE/CMakeLists.txt:
```diff
@@ -1,11 +1,17 @@
+# We use INTERFACE libraries to form the main object library otherwise we would 
+# get circular deps error between ble <-> cordio
+add_library(mbed-ble-cordio INTERFACE)
+add_library(mbed-ble-cordio-ll INTERFACE)
```
Is there a reason `mbed-ble-cordio-ll` is defined here? Should it be in `connectivity/FEATURE_BLE/libraries/TARGET_CORDIO_LL/CMakeLists.txt`.

0xc0170 replied:
I moved them. I wanted to have them in one place with just one comment and let the deps fixed and remove this comment. I spread it now

0xc0170 committed:
CMake connectivity: fix drivers inclusion

Drivers to be the last, and the stack first.

This can be reverted once drivers are either part of the stack (moved) or we get
add/remove libraries from targets (how a target can request additional driver, this
works now due to using target labels).

0xc0170 committed:
CMake connectivity: remove not useful comment

0xc0170 committed:
CMake ble: move cordio add_library to own modules

0xc0170 committed:
CMake ble drivers: fix adding RTOS as library as they require it
